### PR TITLE
CORE-16195 - uniqueness-rpc-processor

### PR DESCRIPTION
**Pull request by driessamyn**
- Integration of a HTTP RPC server into the Uniqueness processor.
- Smoketests include a test for the service outside the scope of a flow.
- Bunch of fixes/changes to the JavalinServer implementation.
- No integration of a client for this service yet, this will come in a subsequent change.

--
NOTE:
Supersedes  https://github.com/corda/corda-runtime-os/pull/4577 
Commits squashed to clean up history, original commits by @LWogan and @mbrkic-r3

### CODE REVIEW

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/StateAndEventSubscriptionImpl.kt [hunk 1]
```diff
@@ -245,6 +261,28 @@ internal class StateAndEventSubscriptionImpl<K : Any, S : Any, E : Any>(
...
+                        .uri(URI.create("http://localhost:7000/uniquenessChecker"))
+                        .headers("Content-Type", "application/octet-stream")
+                        .POST(HttpRequest.BodyPublishers.ofByteArray(payload))
+                        .build()
+                    val response = httpClient.send(request, HttpResponse.BodyHandlers.ofByteArray())
+                    val responseBody: ByteArray = response.body()
+                    @Suppress("UNCHECKED_CAST")
+                    val responseEvent = avroDeserializer.deserialize(responseBody) as? E
+                    log.info("Got a response: $responseEvent")
+                    Record(record.topic, record.key, responseEvent)
```
Comment (emilybowe) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/StateAndEventSubscriptionImpl.kt]: Think this has to be Flow Event topic possibly? Miljenko made a note but may not have pushed it

File: libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt [hunk 2]
```diff
@@ -40,25 +39,19 @@ class JavalinServer(
 
     override fun start(port: Int) {
         if (server != null) {
```
Comment (mbrkic-r3) [inline libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt]: Might use `check()` instead

File: libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt [hunk 3]
```diff
@@ -74,42 +67,60 @@ class JavalinServer(
...
-    override fun stop() {
-        coordinator.updateStatus(LifecycleStatus.DOWN)
+    private fun stopServer() {
         server?.stop()
         server = null
+    }
+
+    private fun restartServer() {
+        // restart server without marking the component down.
+        val port = server?.port()?:throw java.lang.IllegalStateException("Cannot restart a non-existing server")
```
Comment (mbrkic-r3) [inline libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt]: It might be easier to read if `check()` is performed above with `server` instead of `port`

File: libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt [hunk 4]
```diff
@@ -74,42 +67,60 @@ class JavalinServer(
...
+    override fun stop() {
+        coordinator.updateStatus(LifecycleStatus.DOWN)
+        stopServer()
         coordinator.stop()
     }
 
     override fun registerEndpoint(endpoint: Endpoint) {
-        registerEndpointInternal(endpoint)
+        // register immediately when the server has been started
+        if(null != server) registerEndpointInternal(endpoint)
```
Comment (mbrkic-r3) [inline libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:107]: We should see whether we can enforce formatting / reformat code automatically after it's pushed so we have consistent formatting across the codebase (in this file it's not consistent space between `if` and `(`, `server != null` vs `null != server`).

File: libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt [hunk 5]
```diff
@@ -40,25 +39,19 @@ class JavalinServer(
 
     override fun start(port: Int) {
         if (server != null) {
-            throw CordaRuntimeException("The Javalin webserver is already initialized")
+            throw IllegalStateException("The Javalin webserver is already initialized")
```
Comment (mbrkic-r3) [inline libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt]: Tests that check for this exception need to be updated

Comment (driessamyn) [reply libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt]: Y. This is WIP, just trying to get to the point the thing actually starts.
I suspect a lot of tests will need to be updated, and some added as there must have been quite a lot of gaps.

File: libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt (hunk 4 above)
Comment (driessamyn) [reply libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:107]: Simon has previously suggested we introduce an automatic linter. I'm fully supportive.

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/utils/UniquenessAssertions.kt [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import net.corda.test.util.time.AutoTickTestClock
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Assertions.assertInstanceOf
+import java.time.Instant
+
+/**
+ * Uniqueness check related assertions for use in tests
+ */
+@Suppress("SpreadOperator", "TooManyFunctions")
+object UniquenessAssertions {
```
Comment (mbrkic-r3) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/utils/UniquenessAssertions.kt]: We don't need to copy this, we can add it as dependency instead:
`smokeTestImplementation project(":testing:uniqueness:uniqueness-utilities")`

File: applications/workers/workers-smoketest/build.gradle [hunk 7]
```diff
@@ -114,25 +122,29 @@ tasks.register('smokeTest', Test) {
     testClassesDirs = project.sourceSets["smokeTest"].output.classesDirs
     classpath = project.sourceSets["smokeTest"].runtimeClasspath
 
-    def combinedWorker = project.getProperties().getOrDefault("isCombinedWorker",false)
+    // TODO - switch default around so this can be run locally.
+    //  Also refactor property names to remove "health" now that they are more general purpose.
```
Comment (driessamyn) [inline applications/workers/workers-smoketest/build.gradle:126]: This will be addressed in a separate PR (CORE-17202).

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/StateAndEventSubscriptionImpl.kt [hunk 8]
```diff
@@ -245,6 +261,32 @@ internal class StateAndEventSubscriptionImpl<K : Any, S : Any, E : Any>(
 
         commitTimer.recordCallable {
             producer.beginTransaction()
+
+            // HACK for testing only
```
Comment (relyafi) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/StateAndEventSubscriptionImpl.kt]: This looks odd, this isn't test specific code. If this is a temporary change needed to facilitate our current implementation, I'd leave it as a TODO / FIXME with an associated Jira ticket reference that will implement an improved solution.

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+import java.time.Instant
+import java.time.LocalDate
+import java.time.ZoneOffset
+import java.util.UUID
+
+/**
+ * Tests for the UniquenessChecker RPC service
+ */
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class UniquenessChecker {
```
Comment (relyafi) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: This feels a bit too broad a name and doesn't follow the naming precedent of other smoke test classes. I'd suggest something like `UniquenessCheckerRPCTests` or `UniquenessCheckerRPCSmokeTests`.

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        registerStaticMember(bobHoldingId)
+        registerStaticMember(charlieHoldingId)
+        registerStaticMember(notaryHoldingId, NOTARY_SERVICE_X500)
+    }
+
+    @Test
+    fun `when call service with valid payload return idempotently`() {
+        val url = "${System.getProperty("uniquenessWorkerHealthHttp")}api/5.1/uniqueness-checker"
+
+        logger.warn("uniqueness url: $url")
```
Comment (relyafi) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: If this is needed at all, shouldn't it be on `info` not `warn`?

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        assertThat(notaryActualHoldingId).isEqualTo(notaryHoldingId)
+
+        registerStaticMember(aliceHoldingId)
+        registerStaticMember(bobHoldingId)
+        registerStaticMember(charlieHoldingId)
+        registerStaticMember(notaryHoldingId, NOTARY_SERVICE_X500)
+    }
+
+    @Test
+    fun `when call service with valid payload return idempotently`() {
```
Comment (relyafi) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: I'm not sure this test case is proving what it states? We're only sending the request once, so isn't this just proving a trivial success case? I think you want to send twice and verify the exact same response if you want to prove idempotency.

I'd also recommend a second test scenario triggering a failure scenario and proving idempotency here too, since the failure code path through the uniqueness checker is different from the success case. A good way of doing this might be to send a request with a very near time window upper bound, increment the clock past this, and then send the exact same response again. A "new" request would fail the check, but the old one should simply retrieve the original result. It should be possible to manipulate the clock to do this given we're using a test clock.

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+ * events response API.
+ */
+class UniquenessCheckRpcMessageProcessor(
+    private val uniquenessChecker: UniquenessChecker,
+    private val externalEventResponseFactory: ExternalEventResponseFactory,
+    override val requestClass: Class<UniquenessCheckRequestAvro>,
+    override val responseClass: Class<FlowEvent>,
+) : SyncRPCProcessor<UniquenessCheckRequestAvro, FlowEvent> {
+
+    override fun process(request: UniquenessCheckRequestAvro): FlowEvent {
```
Comment (relyafi) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: This implies that we are artificially limiting the RPC interface to process a single request at a time at present. But is there any value in doing this? I appreciate that for the time being, RPC calls will be synchronous and only issue one request at a time, but we could have the server side support a list of requests, and just get the client to invoke with a single request at a time?

This would save us having to revisit this interface and having to worry about backwards compatibility in future.

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        val result = uniquenessChecker.processRequests(listOf(request))
```
Comment (relyafi) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt]: This could be inlined on like 24 as this is only used once.

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+        return result.map { (request, response) ->
+            if (response.result is UniquenessCheckResultUnhandledExceptionAvro) {
+
+                externalEventResponseFactory.platformError(
+                    request.flowExternalEventContext,
+                    (response.result as UniquenessCheckResultUnhandledExceptionAvro).exception
+                )
+
+
```
Comment (relyafi) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt]: Redundant whitespace

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt (hunk 12 above)
Comment (relyafi) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: This function appears to be missing the tracing logic that is present in the (Kafka based) `UniquenessCheckMessageProcessor`. Are we intentionally considering tracing out of scope for now, and if so, should we add a TODO / FIXME with a Jira ticket that will re-introduce this functionality?

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+            } else {
+                externalEventResponseFactory.success(request.flowExternalEventContext, response)
+            }
+        }.first().value!!
```
Comment (relyafi) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt]: If we do adopt my earlier suggestion then this will need to return all results. Otherwise, this code block should probably enforce exactly one result via `single()`, either here or on line 24.

File: components/uniqueness/uniqueness-checker-impl/src/test/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessorTests.kt [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+import net.corda.messaging.api.records.Record
+import net.corda.uniqueness.checker.UniquenessChecker
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import org.mockito.kotlin.any
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.verify
+
+class UniquenessCheckRpcMessageProcessorTests {
```
Comment (relyafi) [inline components/uniqueness/uniqueness-checker-impl/src/test/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessorTests.kt:19]: I'd suggest that the test cases in this suite should include adaptations of those in `MessageBusIntegrationTests`, as this is effectively testing at the same level as those tests. This would give us strong confidence that the new interface behaves identically to the old one. This will likely need doing anyway at the point the Kafka based message bus implementation is removed, as we won't want to lose the associated test cases.

**Review by relyafi — COMMENTED** [review]
Mostly minor comments. As mentioned in one of the specific comments, I'd strongly consider making the server API support a batch of requests, even if the client will only send a single request at a time at present.

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+        assertThat(deserializedExternalEventResponse).isNotNull
+        UniquenessAssertions.assertStandardSuccessResponse(deserializedExternalEventResponse!!, testClock)
+    }
+
+    private val testClock = AutoTickTestClock(Instant.MAX, Duration.ofSeconds(1))
+
+    /**
+     * Returns a random set of bytes
+     */
+    private fun randomBytes(): ByteArray {
```
Comment (mbrkic-r3) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: Might be replaced with `Random.nextBytes(16)`?

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/StateAndEventSubscriptionImpl.kt (hunk 8 above)
Comment (mbrkic-r3) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/StateAndEventSubscriptionImpl.kt]: These changes were not supposed to get merged, they need to be removed.

File: applications/workers/workers-smoketest/build.gradle [hunk 18]
```diff
@@ -114,25 +122,29 @@ tasks.register('smokeTest', Test) {
...
+            project.getProperties().getOrDefault("cryptoWorkerHealthHttp", combinedWorker ? combinedWorkerAddress : "http://localhost:7001/")
     systemProperty "restWorkerHealthHttp",
-            project.getProperties().getOrDefault("restWorkerHealthHttp",combinedWorker ? null : "http://localhost:7002/")
+            project.getProperties().getOrDefault("restWorkerHealthHttp", combinedWorker ? combinedWorkerAddress : "http://localhost:7002/")
     systemProperty "flowWorkerHealthHttp",
-            project.getProperties().getOrDefault("flowWorkerHealthHttp",combinedWorker ? null : "http://localhost:7003/")
+            project.getProperties().getOrDefault("flowWorkerHealthHttp", combinedWorker ? combinedWorkerAddress : "http://localhost:7003/")
+    systemProperty "flowMapperWorkerHealthHttp",
+            project.getProperties().getOrDefault("flowMapperWorkerHealthHttp", combinedWorker ? combinedWorkerAddress : "http://localhost:7004/")
+    systemProperty "flowMapperWorkerHealthHttp",
```
Comment (emilybowe) [inline applications/workers/workers-smoketest/build.gradle]: think this should be `systemProperty verificationWorkerHealthHttp`

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt (hunk 12 above)
Comment (mbrkic-r3) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: Tracing was moved to `SyncRPCSubscriptionImpl.kt`

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+
+        registerStaticMember(aliceHoldingId)
+        registerStaticMember(bobHoldingId)
+        registerStaticMember(charlieHoldingId)
+        registerStaticMember(notaryHoldingId, NOTARY_SERVICE_X500)
+    }
+
+    @Test
+    fun `when call service with valid payload return idempotently`() {
+        val url = "${System.getProperty("uniquenessWorkerHealthHttp")}api/5.1/uniqueness-checker"
```
Comment (mbrkic-r3) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: Should we get the version from `PlatformInfoProvider.kt`?

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt (hunk 11 above)
Comment (mbrkic-r3) [reply applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: The main idea of this test was just to verify that RPC endpoint is working, that we are accepting requests and returning back responses. So I think we should just rename the test. Client side is not finished yet, and whole RPC stuff won't be used until Multisource-Event Mediator is implemented, so I think we don't need to test everything/whole Uniqueness functionality at this time since RPC endpoint won't be used yet.

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt (hunk 12 above)
Comment (mbrkic-r3) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: This could definitely be a future improvement, but I would leave it for another PR, this one is big enough already.

Comment (relyafi) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: Is it a big change? I'd just be worried if we do that then it'll fall into the category of change that will never be considered important enough to do, until we need it, which will then likely require an API version upgrade. Arguably the batch support is less of a change from what we already have.

File: components/uniqueness/uniqueness-checker-impl/src/test/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessorTests.kt (hunk 16 above)
Comment (mbrkic-r3) [reply components/uniqueness/uniqueness-checker-impl/src/test/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessorTests.kt:19]: This is just a unit test (with mocks) for `UniquenessCheckRpcMessageProcessor`. Tests similar to those in `MessageBusIntegrationTests` should probably also be integration tests, e.g. `UniquenessRpcIntegrationTests` (that replaces `MessageBusIntegrationTests` once we are ready for that).

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt (hunk 12 above)
Comment (mbrkic-r3) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: I meant that PR is big enough already to add additional methods and tests. If we just want to replace existing method, it might be a small change, and it might make sense to do it now if that would otherwise be a backward-incompatible change in the future. For initial version of Multi-source Event Mediator we didn't plan to batch RPC requests.

Comment (relyafi) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: I think that would be worth investigating, as we already have batching support available within the existing uniqueness checker functionality so it wouldn't be too onerous to do a quick sanity check the batch interface works. Understand it won't be used by the client in the first instance (i.e. it would always use a batch size of 1).

Comment (emilybowe) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: I've created a ticket to track it https://r3-cev.atlassian.net/browse/CORE-17182

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt (hunk 19 above)
Comment (mbrkic-r3) [reply applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: Do we still use the same port for health and other functionality? I see [this](https://github.com/corda/corda-runtime-os/pull/4470#discussion_r1311559501) comment is still not resolved?

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt (hunk 12 above)
Comment (mbrkic-r3) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: @Omar-awad @driessamyn Please advise.

**Review by ronanbrowne — DISMISSED** [review]
LGTM - build files only reviewed

**Review by dimosr — DISMISSED** [review]
approving for @corda/corda-platform-network-team (only import changes)

**Review by josephzunigadaly — DISMISSED** [review]
Approving for REST owned files only

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt (hunk 19 above)
Comment (emilybowe) [reply applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: for comment 1, discussed offline as this would need to be injected as a service it looks like and might be non-trivial. Smaller improvement added for now to use a constant in utils with ticket created for later https://r3-cev.atlassian.net/browse/CORE-17187

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt (hunk 12 above)
Comment (driessamyn) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: It's valid feedback from @relyafi ,  but I agree that deferring this change is the right thing to do. 
Personally, I think we should have a think about which batching implementations work best. Batching could, for example, happen completely within the Uniqueness Service, i.e. it could batch up requests to process together, rather, or as well as, implementing this on the API.
We should test and validate this I think before doing the implementation.

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt (hunk 19 above)
Comment (driessamyn) [reply applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: > Do we still use the same port for health and other functionality? I see [this](https://github.com/corda/corda-runtime-os/pull/4470#discussion_r1311559501) comment is still not resolved?

Yes, the decision was to prefer a single web server until there is a reason not to.

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt (hunk 11 above)
Comment (driessamyn) [reply applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: Sorry, that is my fault. I initially added `idempotently` to the test name as I thought it would be worth validating this (and add as a regression).

I agree that we should add additional tests though. Let's raise a JIRA so we backfill the test suite with negative and idempotent tests.

**Review by relyafi — APPROVED** [review]
Uniqueness changes LGTM, subject to the topic of batching being revisited as discussed.

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt (hunk 12 above)
Comment (emilybowe) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckRpcMessageProcessor.kt:21]: great thanks. I've linked this thread and the message further down where @relyafi mentions changing single() to returning all results in the [ticket](https://r3-cev.atlassian.net/browse/CORE-17182 ) for this. I'll mark this resolved if no one objects.

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt (hunk 19 above)
Comment (emilybowe) [reply applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/services/UniquenessChecker.kt]: great thanks for clarity on comment 2, marking this thread resolved now

**Review by mbrkic-r3 — APPROVED** [review]
LGTM

**Review by ronanbrowne — APPROVED** [review]

**Review by dimosr — APPROVED** [review]
re-approving for @corda/corda-platform-network-team files

**Review by josephzunigadaly — APPROVED** [review]
Re-approve for REST owned files
